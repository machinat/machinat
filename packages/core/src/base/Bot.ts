import type { SociablyBot, SociablyThread, SociablyNode } from '../types.js';
import type { DispatchResponse } from '../engine/types.js';
import { serviceInterface, serviceProviderClass } from '../service/index.js';

type BaseBotI = SociablyBot<SociablyThread, unknown, unknown>;

const BotPlatformMapI = serviceInterface<BaseBotI>({
  name: 'BotPlatformMap',
  polymorphic: true,
});

/**
 * @category Base
 */
export class BaseBot implements BaseBotI {
  static PlatformMap = BotPlatformMapI;

  private _platformMapping: Map<string, BaseBotI>;

  constructor(platformMapping: Map<string, BaseBotI>) {
    this._platformMapping = platformMapping;
  }

  async render(
    thread: SociablyThread,
    node: SociablyNode
  ): Promise<null | DispatchResponse<unknown, unknown>> {
    const bot = this._platformMapping.get(thread.platform);
    if (!bot) {
      throw new TypeError(
        `thread of platform '${thread.platform}' is not supported`
      );
    }

    return bot.render(thread, node);
  }

  async start(): Promise<void> {} // eslint-disable-line class-methods-use-this
  async stop(): Promise<void> {} // eslint-disable-line class-methods-use-this
}

const BotP = serviceProviderClass({
  lifetime: 'transient',
  deps: [BotPlatformMapI],
})(BaseBot);

type BotP = BaseBot;

export default BotP;
