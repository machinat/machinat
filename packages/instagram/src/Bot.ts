import { AgentSettingsAccessor } from '@sociably/core';
import Engine, { DispatchError } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import Renderer from '@sociably/core/renderer';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { serviceProviderClass } from '@sociably/core/service';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import {
  MetaApiWorker,
  MetaApiChannel,
  MetaApiJob,
  MetaApiResult,
  MetaApiDispatchResponse,
  MetaApiResponseBody,
  MetaRequestApiOptions,
} from '@sociably/meta-api';
import {
  createChatJobs,
  createUploadChatAttachmentJobs,
  MessengerBot,
} from '@sociably/messenger';
import { renderGeneralComponents } from '@sociably/messenger/components';
import { INSTAGRAM } from './constant.js';
import {
  ConfigsI,
  PlatformUtilitiesI,
  AgentSettingsAccessorI,
} from './interface.js';
import InstagramAgent from './Agent.js';
import InstagramChat from './Chat.js';
import type {
  InstagramComponent,
  InstagramSegmentValue,
  InstagramDispatchFrame,
  InstagramMessagingOptions,
  InstagramAgentSettings,
  InstagramThread,
} from './types.js';

type InstagramBotOptions = {
  appId: string;
  appSecret: string;
  graphApiVersion?: string;
  apiBatchRequestInterval?: number;
  agentSettingsAccessor: AgentSettingsAccessor<
    InstagramAgent,
    InstagramAgentSettings
  >;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    MetaApiJob,
    InstagramDispatchFrame,
    MetaApiResult
  >;
};

type UploadAttachmentResult = {
  attachmentId: string;
};

/**
 * InstagramBot render messages and make API call to Instagram platform.
 * @category Provider
 */
export class InstagramBot
  implements
    MessengerBot<MetaApiChannel>,
    SociablyBot<InstagramChat, MetaApiJob, MetaApiResult>
{
  worker: MetaApiWorker;
  engine: Engine<
    InstagramThread,
    InstagramSegmentValue,
    InstagramComponent<unknown>,
    MetaApiJob,
    MetaApiResult
  >;

  platform = INSTAGRAM;

  constructor({
    appId,
    appSecret,
    graphApiVersion = 'v17.0',
    apiBatchRequestInterval = 500,
    agentSettingsAccessor,
    initScope,
    dispatchWrapper,
  }: InstagramBotOptions) {
    const renderer = new Renderer<
      InstagramSegmentValue,
      InstagramComponent<unknown>
    >(INSTAGRAM, renderGeneralComponents);

    const queue = new Queue<MetaApiJob, MetaApiResult>();
    const worker = new MetaApiWorker({
      agentSettingsAccessor,
      appId,
      appSecret,
      graphApiVersion,
      consumeInterval: apiBatchRequestInterval,
    });

    this.engine = new Engine(
      INSTAGRAM,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  async render(
    target: InstagramChat,
    node: SociablyNode
  ): Promise<null | MetaApiDispatchResponse> {
    if (target instanceof InstagramChat) {
      return this.engine.render(
        target,
        node,
        createChatJobs<InstagramAgent, InstagramChat>(target.agent)
      );
    }

    throw new TypeError('invalid rendering target');
  }

  /** Send messages or actions on a chat */
  async message(
    chat: InstagramChat,
    messages: SociablyNode,
    options?: InstagramMessagingOptions
  ): Promise<null | MetaApiDispatchResponse> {
    return this.engine.render(
      chat,
      messages,
      createChatJobs<InstagramAgent, InstagramChat>(chat.agent, options)
    );
  }

  /** Upload a media chat attachment for later use */
  async uploadChatAttachment(
    /** The {@link InstagramAgent} that owns the attachment */
    agentInput: string | InstagramAgent,
    /** An {@link Image}, {@link Audio}, {@link Video} or {@link File} element to be uploaded */ node: SociablyNode
  ): Promise<null | UploadAttachmentResult> {
    const agent =
      typeof agentInput === 'string'
        ? new InstagramAgent(agentInput)
        : agentInput;
    const response = await this.engine.render(
      agent,
      node,
      createUploadChatAttachmentJobs({ platform: 'instagram' })
    );
    const result = response?.results[0].body;
    return result ? { attachmentId: result.attachment_id } : null;
  }

  async requestApi<ResBody extends MetaApiResponseBody>({
    channel: agentInput,
    method = 'GET',
    url,
    params,
    accessToken,
    asApp,
  }: MetaRequestApiOptions<InstagramAgent>): Promise<ResBody> {
    const account =
      typeof agentInput === 'string'
        ? new InstagramAgent(agentInput)
        : agentInput;
    try {
      const { results } = await this.engine.dispatchJobs(account || null, [
        {
          channel: account,
          request: { method, url, params },
          accessToken,
          asApp,
        },
      ]);

      return results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    AgentSettingsAccessorI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { appSecret, appId, apiBatchRequestInterval },
    agentSettingsAccessor,
    moduleUitils,
    platformUtils
  ) =>
    new InstagramBot({
      agentSettingsAccessor,
      appId,
      appSecret,
      apiBatchRequestInterval,
      initScope: moduleUitils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(InstagramBot);

type BotP = InstagramBot;
export default BotP;
