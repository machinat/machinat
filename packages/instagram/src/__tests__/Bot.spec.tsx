import querystring from 'querystring';
import nock from 'nock';
import moxy, { Moxy } from '@moxyjs/moxy';
import Sociably from '@sociably/core';
import _Renderer from '@sociably/core/renderer';
import Queue from '@sociably/core/queue';
import _Engine from '@sociably/core/engine';
import * as MetaApiModule from '@sociably/meta-api';
import { MetaApiWorker as _Worker, MetaApiError } from '@sociably/meta-api';
import { InstagramBot } from '../Bot.js';
import { Image, Expression, TextReply } from '../components/index.js';
import InstagramChat from '../Chat.js';
import InstagramAgent from '../Agent.js';

const Renderer = _Renderer as Moxy<typeof _Renderer>;
const Engine = _Engine as Moxy<typeof _Engine>;
const Worker = _Worker as Moxy<typeof _Worker>;

nock.disableNetConnect();

jest.mock('@sociably/core/engine', () =>
  moxy(jest.requireActual('@sociably/core/engine')),
);

jest.mock('@sociably/core/renderer', () =>
  moxy(jest.requireActual('@sociably/core/renderer')),
);

jest.mock('@sociably/meta-api', () => {
  const module = jest.requireActual<typeof MetaApiModule>('@sociably/meta-api');
  return {
    ...module,
    MetaApiWorker: moxy(module.MetaApiWorker),
  };
});

const initScope = moxy(() => moxy());
const dispatchWrapper = moxy((x) => x);

const accountId = '1234567890';
const pageId = '1111111111';
const accessToken = '_ACCESS_TOKEN_';
const appId = '_APP_ID_';
const appSecret = '_APP_SECRET_';
const username = 'jojodoe666';

const agentSettingsAccessor = {
  getAgentSettings: async () => ({ accountId, pageId, accessToken, username }),
  getAgentSettingsBatch: async () => [
    { accountId, pageId, accessToken, username },
  ],
};

const agent = new InstagramAgent(accountId);

const message = (
  <Expression quickReplies={<TextReply title="Hi!" payload="ðŸ‘‹" />}>
    Hello <b>World!</b>
    <Image url="https://sociably.io/greeting.png" />
  </Expression>
);

let graphApi;
const bodySpy = moxy(() => true);

beforeEach(() => {
  graphApi = nock('https://graph.facebook.com').post('/v17.0/', bodySpy);
  bodySpy.mock.clear();
  Engine.mock.clear();
  Renderer.mock.clear();
  Worker.mock.clear();
});

afterEach(() => {
  nock.cleanAll();
});

describe('.constructor(options)', () => {
  it('assemble core modules', () => {
    const bot = new InstagramBot({
      appId,
      appSecret,
      agentSettingsAccessor,
      initScope,
      dispatchWrapper,
    });

    expect(bot.engine).toBeInstanceOf(Engine);

    expect(Renderer).toHaveBeenCalledTimes(1);
    expect(Renderer).toHaveBeenCalledWith('instagram', expect.any(Function));

    expect(Engine).toHaveBeenCalledTimes(1);
    expect(Engine).toHaveBeenCalledWith(
      'instagram',
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(Worker),
      initScope,
      dispatchWrapper,
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith({
      agentSettingsAccessor,
      appId,
      appSecret,
      graphApiVersion: 'v17.0',
      consumeInterval: 500,
    });
  });

  it('pass options to worker', () => {
    expect(
      new InstagramBot({
        appId,
        appSecret,
        agentSettingsAccessor,
        initScope,
        dispatchWrapper,
        graphApiVersion: 'v8.0',
        apiBatchRequestInterval: 0,
      }),
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith({
      agentSettingsAccessor,
      appId,
      appSecret,
      graphApiVersion: 'v8.0',
      consumeInterval: 0,
    });
  });
});

test('.start() and .stop() start/stop engine', () => {
  const bot = new InstagramBot({
    appId,
    appSecret,
    agentSettingsAccessor,
    initScope,
    dispatchWrapper,
  });

  bot.start();
  expect(bot.engine.start).toHaveBeenCalledTimes(1);

  bot.stop();
  expect(bot.engine.stop).toHaveBeenCalledTimes(1);
});

describe('.message(thread, message, options)', () => {
  const bot = new InstagramBot({ agentSettingsAccessor, appId, appSecret });

  let apiStatus;
  beforeEach(() => {
    const messageResult = {
      code: 200,
      body: JSON.stringify({ message_id: 'xxx', recipient_id: 'xxx' }),
    };
    apiStatus = graphApi.reply(200, [messageResult, messageResult]);
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  const chat = new InstagramChat('1234567890', { id: '9876543210' });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(bot.message(chat, empty)).resolves.toBe(null);
      expect(apiStatus.isDone()).toBe(false);
    }
  });

  it('send messages to me/messages api', async () => {
    const response = await bot.message(chat, message);
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: { message_id: 'xxx', recipient_id: 'xxx' },
      });
    }

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    for (const request of JSON.parse(body.batch)) {
      expect(request.method).toBe('POST');
      expect(request.relative_url).toBe(
        'me/messages?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f',
      );
    }

    expect(body).toMatchSnapshot();
    expect(JSON.parse(body.batch).map((req) => querystring.decode(req.body)))
      .toMatchInlineSnapshot(`
      [
        {
          "message": "{"text":"Hello World!"}",
          "recipient": "{"id":"9876543210"}",
        },
        {
          "message": "{"attachment":{"type":"image","payload":{"url":"https://sociably.io/greeting.png"}},"quick_replies":[{"content_type":"text","title":"Hi!","payload":"ðŸ‘‹"}]}",
          "recipient": "{"id":"9876543210"}",
        },
      ]
    `);

    expect(apiStatus.isDone()).toBe(true);
  });

  test('message options', async () => {
    const response = await bot.message(chat, message, {
      messagingType: 'MESSAGE_TAG',
      tag: 'HUMAN_AGENT',
      notificationType: 'SILENT_PUSH',
      personaId: 'billy17',
    });
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: { message_id: 'xxx', recipient_id: 'xxx' },
      });
    }

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot();
    expect(JSON.parse(body.batch).map((req) => querystring.decode(req.body)))
      .toMatchInlineSnapshot(`
      [
        {
          "message": "{"text":"Hello World!"}",
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": "{"id":"9876543210"}",
          "tag": "HUMAN_AGENT",
        },
        {
          "message": "{"attachment":{"type":"image","payload":{"url":"https://sociably.io/greeting.png"}},"quick_replies":[{"content_type":"text","title":"Hi!","payload":"ðŸ‘‹"}]}",
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": "{"id":"9876543210"}",
          "tag": "HUMAN_AGENT",
        },
      ]
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('.uploadChatAttachment(agent, message)', () => {
  const bot = new InstagramBot({ agentSettingsAccessor, appId, appSecret });

  beforeEach(() => {
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      await expect(bot.uploadChatAttachment(agent, empty)).resolves.toBe(null); // eslint-disable-line no-await-in-loop
    }
  });

  it('call message_attachments api', async () => {
    const apiStatus = graphApi.reply(200, [
      { code: 200, body: JSON.stringify({ attachment_id: 401759795 }) },
    ]);

    await expect(
      bot.uploadChatAttachment(
        agent,
        <Image url="https://sociably.io/trollface.png" />,
      ),
    ).resolves.toEqual({ attachmentId: 401759795 });

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body).toMatchSnapshot();
    const reqest = JSON.parse(body.batch)[0];
    expect(reqest.method).toBe('POST');

    expect(reqest.relative_url).toMatchInlineSnapshot(
      `"me/message_attachments?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f"`,
    );
    expect(querystring.decode(reqest.body)).toMatchInlineSnapshot(`
      {
        "message": "{"attachment":{"type":"image","payload":{"url":"https://sociably.io/trollface.png"}}}",
        "platform": "instagram",
      }
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('.requestApi(options)', () => {
  it('call facebook graph api', async () => {
    const bot = new InstagramBot({ agentSettingsAccessor, appId, appSecret });
    bot.start();

    const apiCall = graphApi.reply(200, [{ code: 200, body: '{"foo":"bar"}' }]);

    await expect(
      bot.requestApi({
        channel: agent,
        method: 'POST',
        url: 'foo',
        params: { bar: 'baz' },
      }),
    ).resolves.toEqual({ foo: 'bar' });

    expect(apiCall.isDone()).toBe(true);
    expect(bodySpy).toHaveBeenCalledTimes(1);
    expect(bodySpy.mock.calls[0].args[0]).toMatchInlineSnapshot(`
      {
        "access_token": "_ACCESS_TOKEN_",
        "batch": "[{"method":"POST","relative_url":"foo?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f","body":"bar=baz","omit_response_on_success":false}]",
      }
    `);
  });

  test('with accessToken option', async () => {
    const bot = new InstagramBot({ agentSettingsAccessor, appId, appSecret });
    await bot.start();

    graphApi.reply(200, [{ code: 200, body: '{"foo":"bar"}' }]);
    await expect(
      bot.requestApi({
        channel: agent,
        method: 'POST',
        url: 'foo',
        params: { bar: 'baz' },
        accessToken: '_MY_ACCESS_TOKEN_',
      }),
    ).resolves.toEqual({ foo: 'bar' });

    expect(bodySpy.mock.calls[0].args[0]).toMatchInlineSnapshot(`
      {
        "access_token": "_MY_ACCESS_TOKEN_",
        "batch": "[{"method":"POST","relative_url":"foo?access_token=_MY_ACCESS_TOKEN_&appsecret_proof=a2d4a2635b030a35c39f30231ffe768f18af5b034c7be4e095c48495341db374","body":"bar=baz","omit_response_on_success":false}]",
      }
    `);
  });

  test('with asApp option', async () => {
    const bot = new InstagramBot({ agentSettingsAccessor, appId, appSecret });
    await bot.start();

    graphApi.reply(200, [{ code: 200, body: '{"foo":"bar"}' }]);
    await expect(
      bot.requestApi({
        channel: agent,
        method: 'POST',
        url: 'foo',
        params: { bar: 'baz' },
        asApp: true,
      }),
    ).resolves.toEqual({ foo: 'bar' });

    expect(bodySpy.mock.calls[0].args[0]).toMatchInlineSnapshot(`
      {
        "access_token": "_APP_ID_|_APP_SECRET_",
        "batch": "[{"method":"POST","relative_url":"foo?access_token=_APP_ID_%7C_APP_SECRET_&appsecret_proof=b345404c0883034d5e07120293737f58fc94529813075bd0ddb69dc0f0cd4e1b","body":"bar=baz","omit_response_on_success":false}]",
      }
    `);
  });

  it('throw MetaApiError if api call fail', async () => {
    const bot = new InstagramBot({ agentSettingsAccessor, appId, appSecret });
    bot.start();

    const apiCall = graphApi.reply(200, [
      {
        code: 400,
        body: JSON.stringify({
          error: {
            message: 'bad',
            type: 'devil',
            code: 444,
            error_subcode: 666,
            fbtrace_id: 'xxxx',
          },
        }),
      },
    ]);

    try {
      await bot.requestApi({
        channel: agent,
        method: 'POST',
        url: 'foo',
        params: { bar: 'baz' },
      });
      expect('should not be here').toBeFalsy();
    } catch (err) {
      expect(err).toBeInstanceOf(MetaApiError);
      expect(err.code).toBe(444);
      expect(err.message).toBe('bad');
    }

    expect(apiCall.isDone()).toBe(true);
  });
});
