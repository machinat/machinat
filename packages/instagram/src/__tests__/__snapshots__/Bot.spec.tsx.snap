// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.message(thread, message, options) message options 1`] = `
{
  "jobs": [
    {
      "assetTag": undefined,
      "channel": InstagramPage {
        "$$typeofChannel": true,
        "id": "1234567890",
        "platform": "instagram",
        "username": undefined,
      },
      "file": undefined,
      "key": "ig.1234567890.9876543210",
      "request": {
        "method": "POST",
        "params": {
          "message": {
            "text": "Hello World!",
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": {
            "id": "9876543210",
          },
          "tag": "HUMAN_AGENT",
        },
        "url": "me/messages",
      },
    },
    {
      "assetTag": undefined,
      "channel": InstagramPage {
        "$$typeofChannel": true,
        "id": "1234567890",
        "platform": "instagram",
        "username": undefined,
      },
      "file": undefined,
      "key": "ig.1234567890.9876543210",
      "request": {
        "method": "POST",
        "params": {
          "message": {
            "attachment": {
              "payload": {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": [
              {
                "content_type": "text",
                "image_url": undefined,
                "payload": "ðŸ‘‹",
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": "MESSAGE_TAG",
          "notification_type": "SILENT_PUSH",
          "persona_id": "billy17",
          "recipient": {
            "id": "9876543210",
          },
          "tag": "HUMAN_AGENT",
        },
        "url": "me/messages",
      },
    },
  ],
  "results": [
    {
      "body": {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    {
      "body": {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": [
    {
      "payload": [
        {
          "assetTag": undefined,
          "channel": InstagramPage {
            "$$typeofChannel": true,
            "id": "1234567890",
            "platform": "instagram",
            "username": undefined,
          },
          "file": undefined,
          "key": "ig.1234567890.9876543210",
          "request": {
            "method": "POST",
            "params": {
              "message": {
                "text": "Hello World!",
              },
              "messaging_type": "MESSAGE_TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": {
                "id": "9876543210",
              },
              "tag": "HUMAN_AGENT",
            },
            "url": "me/messages",
          },
        },
        {
          "assetTag": undefined,
          "channel": InstagramPage {
            "$$typeofChannel": true,
            "id": "1234567890",
            "platform": "instagram",
            "username": undefined,
          },
          "file": undefined,
          "key": "ig.1234567890.9876543210",
          "request": {
            "method": "POST",
            "params": {
              "message": {
                "attachment": {
                  "payload": {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": [
                  {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": "ðŸ‘‹",
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": "MESSAGE_TAG",
              "notification_type": "SILENT_PUSH",
              "persona_id": "billy17",
              "recipient": {
                "id": "9876543210",
              },
              "tag": "HUMAN_AGENT",
            },
            "url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`.message(thread, message, options) message options 2`] = `
{
  "access_token": "_ACCESS_TOKEN_",
  "batch": "[{"method":"POST","relative_url":"me/messages?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f","body":"message=%7B%22text%22%3A%22Hello%20World!%22%7D&messaging_type=MESSAGE_TAG&tag=HUMAN_AGENT&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%229876543210%22%7D","name":"ig.1234567890.9876543210-1","omit_response_on_success":false},{"method":"POST","relative_url":"me/messages?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f","body":"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%2C%22payload%22%3A%22%F0%9F%91%8B%22%7D%5D%7D&messaging_type=MESSAGE_TAG&tag=HUMAN_AGENT&notification_type=SILENT_PUSH&persona_id=billy17&recipient=%7B%22id%22%3A%229876543210%22%7D","depends_on":"ig.1234567890.9876543210-1","name":"ig.1234567890.9876543210-2","omit_response_on_success":false}]",
}
`;

exports[`.message(thread, message, options) send messages to me/messages api 1`] = `
{
  "jobs": [
    {
      "assetTag": undefined,
      "channel": InstagramPage {
        "$$typeofChannel": true,
        "id": "1234567890",
        "platform": "instagram",
        "username": undefined,
      },
      "file": undefined,
      "key": "ig.1234567890.9876543210",
      "request": {
        "method": "POST",
        "params": {
          "message": {
            "text": "Hello World!",
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": {
            "id": "9876543210",
          },
          "tag": undefined,
        },
        "url": "me/messages",
      },
    },
    {
      "assetTag": undefined,
      "channel": InstagramPage {
        "$$typeofChannel": true,
        "id": "1234567890",
        "platform": "instagram",
        "username": undefined,
      },
      "file": undefined,
      "key": "ig.1234567890.9876543210",
      "request": {
        "method": "POST",
        "params": {
          "message": {
            "attachment": {
              "payload": {
                "attachment_id": undefined,
                "is_reusable": undefined,
                "url": "https://sociably.io/greeting.png",
              },
              "type": "image",
            },
            "metadata": undefined,
            "quick_replies": [
              {
                "content_type": "text",
                "image_url": undefined,
                "payload": "ðŸ‘‹",
                "title": "Hi!",
              },
            ],
          },
          "messaging_type": undefined,
          "notification_type": undefined,
          "persona_id": undefined,
          "recipient": {
            "id": "9876543210",
          },
          "tag": undefined,
        },
        "url": "me/messages",
      },
    },
  ],
  "results": [
    {
      "body": {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
    {
      "body": {
        "message_id": "xxx",
        "recipient_id": "xxx",
      },
      "code": 200,
    },
  ],
  "tasks": [
    {
      "payload": [
        {
          "assetTag": undefined,
          "channel": InstagramPage {
            "$$typeofChannel": true,
            "id": "1234567890",
            "platform": "instagram",
            "username": undefined,
          },
          "file": undefined,
          "key": "ig.1234567890.9876543210",
          "request": {
            "method": "POST",
            "params": {
              "message": {
                "text": "Hello World!",
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": {
                "id": "9876543210",
              },
              "tag": undefined,
            },
            "url": "me/messages",
          },
        },
        {
          "assetTag": undefined,
          "channel": InstagramPage {
            "$$typeofChannel": true,
            "id": "1234567890",
            "platform": "instagram",
            "username": undefined,
          },
          "file": undefined,
          "key": "ig.1234567890.9876543210",
          "request": {
            "method": "POST",
            "params": {
              "message": {
                "attachment": {
                  "payload": {
                    "attachment_id": undefined,
                    "is_reusable": undefined,
                    "url": "https://sociably.io/greeting.png",
                  },
                  "type": "image",
                },
                "metadata": undefined,
                "quick_replies": [
                  {
                    "content_type": "text",
                    "image_url": undefined,
                    "payload": "ðŸ‘‹",
                    "title": "Hi!",
                  },
                ],
              },
              "messaging_type": undefined,
              "notification_type": undefined,
              "persona_id": undefined,
              "recipient": {
                "id": "9876543210",
              },
              "tag": undefined,
            },
            "url": "me/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`.message(thread, message, options) send messages to me/messages api 2`] = `
{
  "access_token": "_ACCESS_TOKEN_",
  "batch": "[{"method":"POST","relative_url":"me/messages?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f","body":"message=%7B%22text%22%3A%22Hello%20World!%22%7D&recipient=%7B%22id%22%3A%229876543210%22%7D","name":"ig.1234567890.9876543210-1","omit_response_on_success":false},{"method":"POST","relative_url":"me/messages?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f","body":"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Fgreeting.png%22%7D%7D%2C%22quick_replies%22%3A%5B%7B%22content_type%22%3A%22text%22%2C%22title%22%3A%22Hi!%22%2C%22payload%22%3A%22%F0%9F%91%8B%22%7D%5D%7D&recipient=%7B%22id%22%3A%229876543210%22%7D","depends_on":"ig.1234567890.9876543210-1","name":"ig.1234567890.9876543210-2","omit_response_on_success":false}]",
}
`;

exports[`.uploadChatAttachment(page, message) call message_attachments api 1`] = `
{
  "access_token": "_ACCESS_TOKEN_",
  "batch": "[{"method":"POST","relative_url":"me/message_attachments?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f","body":"message=%7B%22attachment%22%3A%7B%22type%22%3A%22image%22%2C%22payload%22%3A%7B%22url%22%3A%22https%3A%2F%2Fsociably.io%2Ftrollface.png%22%7D%7D%7D","omit_response_on_success":false}]",
}
`;
